src/lib/context.js

```diff
@@ -57,39 +57,48 @@ export class Context {
           () => langs);
   }
 
-  _resolve(langs, id, args, formatter) {
+  _resolve(langs, keys, formatter, prevResolved) {
     const lang = langs[0];
 
     if (!lang) {
-      this._env.emit('notfounderror', new L10nError(
-        '"' + id + '"' + ' not found in any language', id), this);
-      if (formatter === this._formatEntity) {
-        return { value: id, attrs: null };
-      } else {
-        return id;
-      }
+      return reportMissing.call(this, keys, formatter, prevResolved);
     }
 
-    const entity = this._getEntity(lang, id);
+    let hasUnresolved = false;
+
+    const resolved = keys.map((key, i) => {
+      if (prevResolved && prevResolved[i] !== undefined) {
+        return prevResolved[i];
+      }
+      const [id, args] = Array.isArray(key) ?
+        key : [key, undefined];
+      const entity = this._getEntity(lang, id);
 
-    if (entity) {
-      return Promise.resolve(
-        formatter.call(this, lang, args, entity, id));
-    } else {
-      this._env.emit('notfounderror', new L10nError(
-        '"' + id + '"' + ' not found in ' + lang.code, id, lang), this);
+      if (entity) {
+        return formatter.call(this, lang, args, entity, id);
+      }
+
+      this._env.emit('notfounderror',
+        new L10nError('"' + id + '"' + ' not found in ' + lang.code,
+          id, lang), this);
+      hasUnresolved = true;
+      return undefined;
+    });
+
+    if (!hasUnresolved) {
+      return Promise.resolve(resolved);
     }
 
-    return this.fetch(langs.slice(1)).then(
-      nextLangs => this._resolve(nextLangs, id, args, formatter));
+    return this.fetch(langs.slice(1)).then(nextLangs =>
+      this._resolve(nextLangs, keys, formatter, resolved));
```

Comment:
Nit: for one-liners, I try to keep the argument and the arrow in the same line as the expression.  (Like it used to be here.)